### PR TITLE
Concept: Custom TownBlockTypes

### LINKED ISSUE #370: Customizable Permission Manager Plots
Originally reported on Google Code with ID 334

```
Zalastri--:
Okay, the feature would be "The ability to define custom plot types, and permissions
inheritance when inside those plots"
That way, in the "plots cfg" section I could say "give plot" "permissions: commandbook.give"
Then, anyone who buys a give plot can use commandbook's give in that plot
while they're there.
That way any plugin can be tied in to towny easily
you could set custom prices per-plot in the "plot cfg"
So price, permissions, and name
and maybe requires: nation/town
That way things like the chest shops being integrated in to a shop plot can be done
easily, by the server admin

Irc Logs of the idea: http://pastie.org/2630396

```

Reported by `zrenfire` on 2011-10-03 05:34:17

### LINKED ISSUE #5072: Suggestion: TownyAPI - Allow for custom TownBlockType
**Please explain your feature request to the best of your abilities:**
Essentially this feature is requesting for TownyAPI to allow customized [TownBlockType](https://javadoc.jitpack.io/com/github/TownyAdvanced/Towny/0.97.0.10/javadoc/com/palmergames/bukkit/towny/object/TownBlockType.html). This will allow developers to create custom TownBlockType's for their plugins and make it easy for plugins to add additional features to towns on the physical layout of towns.

Currently the only available TownBlockTypes are
- ARENA
- BANK
- COMMERCIAL
- EMBASSY
- FARM
- INN
- JAIL
- RESIDENTIAL
- SPLEEF
- WILDS

With an expansion in this area, for example; I can make a custom TownBlockType called "PORT" and add my plugins features to this custom type.

### PR DESCRIPTION
**Pull request by GNosii**
#### Description: 

Allow support for external plugins to register new town block types using the API.

**This PR is not production-ready**
____
#### New Nodes/Commands/ConfigOptions: 

`plugin.allow_external_plot_types` Allow or disable custom plot types

____
#### Relevant Towny Issue ticket:

Closes #5072 

____
- [ ] I have tested this pull request for defects on a server. 

By making this pull request, I represent that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the TownyAdvanced organization has the copyright to use and modify my contribution under the Towny [License](https://github.com/LlmDl/Towny/blob/master/LICENSE.md) for perpetuity.

### CODE REVIEW

**Comment by GNosii** [comment]
After searching for issues related to this in the repository I found that by merging this PR, we can (possibly) close #370 (which dates back to 2011)

**Comment by LlmDl** [comment]
Depends on whether the plot is made to give out the permission nodes or not, which is what #370 calls for. A long while back it was discussed that mucking with bukkit's permissions on plot-change could get busy.

**Comment by GNosii** [comment]
So, this PR is pretty much ready (in theory), I just need to test if this actually works

**Comment by GNosii** [comment]
And also, I need to implement the logic for the plugin.allow_external_plot_types setting, but I'll debug the actual ""API"" first

File: src/com/palmergames/bukkit/towny/listeners/TownyTypeListener.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+/**
+ * Listener for the internal Towny Custom Plot Type API
+ * @author gnosii
+ */
+public class TownyTypeListener implements Listener {
+	
+	@EventHandler
+	public void onTownyBuild(TownyBuildEvent event) {
+		TownBlock tb = event.getTownBlock();
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/listeners/TownyTypeListener.java:18]: All of these events are fired in the wilderness as well as in Towns, so you will need to check if the event is in the wilderness (and return) before you start acting upon the tb variable here.

Comment (GNosii) [reply src/com/palmergames/bukkit/towny/listeners/TownyTypeListener.java]: Resolved in commit 66f5a48

**Comment by GNosii** [comment]
The PR is also missing the SQL data source logic which I'll try to implement over the following hours

**Comment by Warriorrrr** [comment]
Closing this in favor of #5388.

**Pull request closed without merging**